src/Symfony/Component/HttpKernel/ResettingRequestStack.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * Resets services on second master requests.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class ResettingRequestStack extends RequestStack
+{
+    private $resetNeeded = false;
+    private $resettableServices;
+    private $resetMethods;
+
+    public function __construct(\Traversable $resettableServices, array $resetMethods)
```

Comment:
previously the services were instantiated lazily together with the listener on dispatch of `kernel.terminate`, right? Seems now the services would be instantiated before the first master request is handled already, even though they are not necessarily all used during the request.

Means now its affecting performance maybe a bit more than before? I mean before at least it was done after the response was streamed back to the client. Now it will delay handling the response in some cases maybe.

Maybe its possible to do completely really lazy only if they are actually reset?

Reply:
The iterator yields only instantiated services, so all is fine.